### PR TITLE
Merge 01 (base) into muc-cleaner

### PR DESCRIPTION
**Pull request by roperto**
At this point `muc-cleaner` branch is at `master`. I will make PR's into that so you can review, until `muc-cleaner` is ready to be merged into `master`.

### CODE REVIEW

File: cleaner/muc/classes/dml/muc_config_db.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ */
+class muc_config_db {
+    const TABLE_NAME = 'cleaner_muc_configs';
+
+    public static function save($wwwroot, $configuration) {
+        global $DB;
+
+        static::delete($wwwroot);
+
+        // The wwwwroot is base64 encoded to prevent being washed during cleanup.
```
Comment (brendanheywood) [inline cleaner/muc/classes/dml/muc_config_db.php]: 4 w's instead of 3 :)

File: cleaner/muc/lang/en/cleaner_muc.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * @author      Daniel Thee Roperto <daniel.roperto@catalyst-au.net>
+ * @copyright   2017 Catalyst IT Australia {@link http://www.catalyst-au.net}
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @var $plugin stdClass
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+$string['downloader_link'] = 'Download MUC Config file.';
+$string['downloader_in_production'] = 'Sorry, downloading the MUC Config file is not allowed in production environment.';
```
Comment (brendanheywood) [inline cleaner/muc/lang/en/cleaner_muc.php]: this isn't used? just delete it?

File: cleaner/muc/classes/output/index_renderer.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        $renderer = $PAGE->get_renderer('core', null, RENDERER_TARGET_GENERAL);
+
+        return $renderer->heading(get_string('setting_downloader', 'cleaner_muc')) .
+               $this->render_download_link();
+    }
+
+    private function render_download_link() {
+        $url = new moodle_url('/local/datacleaner/cleaner/muc/download.php', ['sesskey' => sesskey()]);
+        $filename = controller::get_download_filename();
+        return '<a download="' . $filename . '" href="' . $url . '">' .
```
Comment (brendanheywood) [inline cleaner/muc/classes/output/index_renderer.php]: I was expecting a Content-Disposition: header, this download attribute is cool :)

File: cleaner/muc/classes/dml/muc_config_db.php (hunk 1 above)
Comment (roperto) [inline cleaner/muc/classes/dml/muc_config_db.php]: Nice catch :-)

File: cleaner/muc/lang/en/cleaner_muc.php (hunk 2 above)
Comment (roperto) [inline cleaner/muc/lang/en/cleaner_muc.php]: Not anymore. Deleted.

File: cleaner/muc/classes/output/index_renderer.php (hunk 3 above)
Comment (roperto) [inline cleaner/muc/classes/output/index_renderer.php]: It saves to messing with headers, right?

Chrome is smart enough to detect if there were errors, so you don't get a corrupt file

image

I am not sure about other browsers though.

**Pull request merged**